### DISCUSSION TITLE
Could the neper output the 'ipf triangle'?

### DISCUSSION THREAD

**Discussion by Yuanguo1994** [neperfepx/neper#239] [Q&A] [answered]
Hi Romain,

I want to know whether the neper could output the 'ipf triangle' after I colored the tessellation by using command 'datacellcol 'ori' -datacellcolscheme 'ipf(z)'' ? If it could output the 'ipf triangle', what's the command I should type in？

Thanks

**Comment by rquey** [comment, marked as answer]
To generate the IPF color key, proceed as follows (requires version `4.2.1-4`):
```
$ neper -T -n 1 -domain "stdtriangle(20)" -dim 2 -o stdtriangle
$ neper -M stdtriangle.tess -cl 0.02 -statnode col_stdtriangle
$ neper -V stdtriangle.msh -datanodecol "col:file(stdtriangle.stnode)" -dataeltcol from_nodes -dataelt2dedgerad 0 -dataelt1drad 0.001 -showelt1d all -print tmp
$ convert -pointsize 40 -draw "text 287,740 '[001]' text 830,740 '[011]' text 767,195 '[111]'" tmp.png stdtriangle.png
```
![image](https://user-images.githubusercontent.com/2659376/125760071-bd67f357-0e17-4073-9363-8cc6dab80d65.png)

**Comment by yguilhem** [comment]
Very nice!
Just tested it with the brand new version (4.2.1-5), it seems that there is a typo in the second command line:
```
$ neper -M stdtriangle.tess -cl 0.02 -statnode col:stdtriangle
========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.2.1-5
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 8 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2021, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/guilhem/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] stdtriangle.tess -cl 0.02 -statnode col:stdtriangle
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `stdtriangle.tess'...
Info   :     [i] Parsed file `stdtriangle.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.02) 
Info   :   - 0D meshing... 100%
Info   :   - 1D meshing... 100%
Info   :   - 2D meshing... 100% (0.94|0.94/100%| 0%| 0%)
Info   : Searching nsets and fasets...
Info   : Writing mesh results...
Info   :   - Preparing mesh...
Info   :   - Mesh properties:
Info   :     > Node number:      295
Info   :     > Elt  number:      521
Info   :     > Mesh area:    0.079
Info   :   - Writing mesh...
Info   :     [o] Writing file `stdtriangle.msh'...
Info   :     [o] Wrote file `stdtriangle.msh'.
Info   : Writing mesh statistics...
Info   :     [o] Writing file `stdtriangle.stnode'...
Error  : Expression `col:stdtriangle' could not be processed.
Aborted
```
But using underscore works fine (see commit [1067c6](https://github.com/rquey/neper/commit/1067c6ec0bc0d3e0f27620ad969f19192f46da5f)):
```
$ neper -M stdtriangle.tess -cl 0.02 -statnode col_stdtriangle

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.2.1-5
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 8 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2021, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/guilhem/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] stdtriangle.tess -cl 0.02 -statnode col_stdtriangle
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `stdtriangle.tess'...
Info   :     [i] Parsed file `stdtriangle.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.02) 
Info   :   - 0D meshing... 100%
Info   :   - 1D meshing... 100%
Info   :   - 2D meshing... 100% (0.94|0.94/100%| 0%| 0%)
Info   : Searching nsets and fasets...
Info   : Writing mesh results...
Info   :   - Preparing mesh...
Info   :   - Mesh properties:
Info   :     > Node number:      295
Info   :     > Elt  number:      521
Info   :     > Mesh area:    0.079
Info   :   - Writing mesh...
Info   :     [o] Writing file `stdtriangle.msh'...
Info   :     [o] Wrote file `stdtriangle.msh'.
Info   : Writing mesh statistics...
Info   :     [o] Writing file `stdtriangle.stnode'...
Info   :     [o] Wrote file `stdtriangle.stnode'.
Info   : Elapsed time: 2.003 secs.
========================================================================
```

**Reply by rquey** [reply]
Thanks @yguilhem , you are right. I corrected my answer.

**Reply by xshang93** [reply]
Tested with 4.2.1-7 on OSX, got below. Not sure what I did wrong?
![stdtriangle](https://user-images.githubusercontent.com/86078353/140259147-9b71b02e-f2c4-44f8-bb7c-5ca4529f91f6.png)

**Reply by rquey** [reply]
Same version - I cannot reproduce this bug (on Linux).

**Comment by qdolle15** [comment]
Hi everyone !

I tried to do the same on my computer (Ubuntu 20.04 - Neper 4.2.1-12)  and it gives me the following error at this stage :

neper -M stdtriangle.tess -cl 0.02 -statnode col_stdtriangle

Can someone help me ? 

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.2.1-12
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 8 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2021, and GNU GPL'd, by Romain Quey.
Info   : No initialization file found (`/home/dolle/.neperrc').

Info   : ---------------------------------------------------------------
Info   : MODULE  -M loaded with arguments:
Info   : [ini file] (none)
Info   : [com line] stdtriangle.tess -cl 0.02 -statnode col_stdtriangle

Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   :   - Reading arguments...
Info   : Loading input data...
Info   :   - Loading tessellation...
Info   :     [i] Parsing file `stdtriangle.tess'...
Info   :     [i] Parsed file `stdtriangle.tess'.
Info   : Meshing...
Info   :   - Preparing... (cl = 0.02) 
Info   :   - 0D meshing... 100%
Info   :   - 1D meshing... 100%
Info   :   - 2D meshing...   0% (0|0/ 0%| 0%| 0%)Error  :     > Meshing of face 1 failed
Abandon (core dumped)

**Reply by xshang93** [reply]
Did you check if `Neper -M` works? To me this looks like a `Neper -M` issue, possibly Gmsh not working.

**Reply by qdolle15** [reply]
Yes you're right, 

When I use M module on a simple example, some "tmpXXXX-X.geo" are created while having the same error message.

What can I do with Gmsh to make it work ?

**Reply by xshang93** [reply]
> Yes you're right,
> 
> When I use M module on a simple example, some "tmpXXXX-X.geo" are created while having the same error message.
> 
> What can I do with Gmsh to make it work ?

it's hard to diagnose, but you can check it here #493 

Otherwise, on mac if I use 8 treads I would have this problem. Seeing that you are on Ubuntu this should not be the case, but it doesn't hurt to try `export OMP_NUM_THREADS=1` before your commands so to only use one thread and see if this fix the problem for you.

**Reply by rquey** [reply]
@qdolle15, please make sure that you installed Gmsh properly (all tests should pass when you run `ctest`). If the problem persists, please open a new issue, as this issue is not related to your problem.

**Reply by qdolle15** [reply]
Ok thanks for your answers !
